File: fixtures/phase5-multilang/go/safe.go
package phase5

import (
    "os/exec"
    "github.com/gin-gonic/gin"
)

func safeRoutes(router *gin.Engine) {
    router.POST("/safe", safeCommand)
}

func safeCommand(context *gin.Context) {
    if !authorized(context) {
        return
    }
    input := context.Query("tool")
    safe := allowlist(input)
    exec.Command("tool", safe).Run()
}

File: fixtures/phase5-multilang/rust/src/safe.rs
use axum::{Router, routing::post};
use std::process::Command;

fn routes() -> Router {
    Router::new().route("/safe", post(safe_command))
}

async fn safe_command(input: String) {
    if !authorized() {
        return;
    }
    let safe = allowlist(input);
    Command::new("tool").arg(safe).status();
}

unsafe fn raw_pointer_boundary(pointer: *const u8) -> u8 {
    unsafe { *pointer }
}
